ocaml/xapi/xapi_psr.ml

```diff
@@ -145,11 +167,214 @@ functor
         )
 
     let start pool_secrets ~master ~members =
-      match Impl.retrieve_checkpoint () |> Option.map checkpoint_of_string with
-      | None | Some No_checkpoint ->
-          go pool_secrets master members No_checkpoint
-      | Some checkpoint ->
-          go (Impl.retrieve ()) master members checkpoint
+      try
+        match
+          Impl.retrieve_checkpoint () |> Option.map checkpoint_of_string
+        with
+        | None | Some No_checkpoint ->
+            go pool_secrets master members No_checkpoint
+        | Some checkpoint ->
+            go (Impl.retrieve ()) master members checkpoint
+      with e ->
+        (* _in theory_ save_checkpoint or backup could fail, so
+           catch that here. however we don't expect this to happen *)
+        D.error "PSR.start: unexpected error: %s" (Printexc.to_string e) ;
+        raise
+          Api_errors.(
+            Server_error (internal_error, ["PSR.start: unexpected error"]))
+  end
+
+let perm = 0o640
+
+let cleanup_internal ~additional_files_to_remove () =
+  if List.length !Xapi_globs.pool_secrets <= 1 then
+    D.info "xapi_psr.ml:cleanup_internal: already cleaned up"
+  else
+    (* if we don't remove pool secret backups, they will be loaded into memory when xapi starts *)
+    let files_to_remove =
+      List.append additional_files_to_remove
+        [old_pool_secret_backup_path; new_pool_secret_backup_path]
+    in
+    List.iter
+      (fun path -> if Sys.file_exists path then Sys.remove path)
+      files_to_remove ;
+    (* psr done, so stop accepting old pool secret *)
+    Xapi_globs.pool_secrets := [Xapi_globs.pool_secret ()]
+
+module Impl =
+functor
+  (Ctx : sig
+     val __context : Context.t
+   end)
+  ->
+  struct
+    open Ctx
+
+    type pool_secret = SecretString.t
+
+    type pool_secrets = pool_secret * pool_secret
+
+    type host = API.ref_host
+
+    let save_checkpoint checkpoint =
+      Unixext.atomic_write_to_file checkpoint_path perm (fun fd ->
+          let (_ : int) =
+            Unix.write_substring fd checkpoint 0 (String.length checkpoint)
+          in
+          ())
+
+    let retrieve_checkpoint () =
+      if not (Sys.file_exists checkpoint_path) then
+        None
+      else
+        let checkpoints = Unixext.read_lines ~path:checkpoint_path in
+        match checkpoints with
```

Comment:
I think it's better to do the `Unixext.read_lines ~path:checkpoint_path ` inside the match and also match exceptions. This could include handling that the file does not exist. There is an unlikely race here between the test earlier and the read.

Reply:
The exception thrown when the file is not found looks like this:
```
open_in "doesn't exist";;
Exception: Sys_error "doesn't exist: No such file or directory".
```

Reply:
thanks! i also didn't know you could match on exceptions in match blocks

Reply:
It got added in ocaml 4.02, so it's available for all backports as well :)